### PR TITLE
#154 add RandomBrowser annotation

### PR DESCRIPTION
**Pull request by oomelianchuk**

### CODE REVIEW

File: src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/RandomBrowser.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import java.lang.annotation.Target;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target(
+{
+  ElementType.TYPE, ElementType.METHOD
+})
+public @interface RandomBrowser
+{
+    int value() default 1;
```
Comment (occupant23) [inline src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/RandomBrowser.java:15]: We should think about the edge cases and test for them in the implementation of the `BrowserStatement` class.

Add a general JavaDoc to the annotation class that explains the usage and states that every value below 1 will deactivate this annotation (similar) to `DataSet`

Implement test cases that cover this.

File: src/main/java/com/xceptance/neodymium/module/statement/browser/BrowserStatement.java [hunk 2]
```diff
@@ -285,6 +289,17 @@ private boolean canReuse(boolean preventReuse, WebDriverStateContainer webDriver
...
+        if (!classRandomBrowserAnnotation.isEmpty() && methodBrowserAnnotations.isEmpty())
+        {
+            Collections.shuffle(browserAnnotations, Neodymium.getRandom());
+            browserAnnotations = browserAnnotations.subList(0, classRandomBrowserAnnotation.get(0).value());
+        }
+        if (!methodRandomBrowserAnnotations.isEmpty())
+        {
+            Collections.shuffle(browserAnnotations, Neodymium.getRandom());
+            browserAnnotations = browserAnnotations.subList(0, methodRandomBrowserAnnotations.get(0).value());
+        }
```
Comment (occupant23) [inline src/main/java/com/xceptance/neodymium/module/statement/browser/BrowserStatement.java:301]: Add a check if more browsers then available should be chosen.

File: src/test/java/com/xceptance/neodymium/testclasses/browser/RandomnessOfRandomBrowser.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+@Browser("browser31")
+@Browser("browser32")
+@Browser("browser33")
+@Browser("browser34")
+@Browser("browser35")
+@Browser("browser36")
+@Browser("browser37")
+@Browser("browser38")
+@Browser("browser39")
+@Browser("browser40")
```
Comment (occupant23) [inline src/test/java/com/xceptance/neodymium/testclasses/browser/RandomnessOfRandomBrowser.java:54]: pick 10 out of 100 like for random data sets

**Review by occupant23 — CHANGES_REQUESTED** [review]
Please adjust and discuss the comments.

File: src/main/java/com/xceptance/neodymium/module/statement/browser/BrowserStatement.java [hunk 4]
```diff
@@ -285,6 +289,17 @@ private boolean canReuse(boolean preventReuse, WebDriverStateContainer webDriver
             browserAnnotations.addAll(classBrowserAnnotations);
         }
 
+        if (!classRandomBrowserAnnotation.isEmpty() && methodBrowserAnnotations.isEmpty())
```
[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/main/java/com/xceptance/neodymium/module/statement/browser/BrowserStatement.java]: Replacement of those two statements with `if(){}else if(){}` structure and removal of one part of the first condition would change the expected behavior.

1. Removal of one part of the first condition:
Using the `@RandomBrowsers` annotation, I'd expect that it's possible to overwrite the `@RandomBrowsers` from the class level with simply `@Browser` annotation on the method level. This is why we need to check, whether the method has the `@Browser`. 

2. `else if` instead of two `if`s:
Analog to most of the other annotations in the Neodymium, the `@RandomBrowsers` annotation of the class level should be overwritable on the method level. Using `else if`  destroys this principle because if the annotation is present for class, the `else` wouldn't allow us to check if it should be overwritten on the method level.

Please let me know if the described behavior is not the desired one or if I got your comment wrong.

File: src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/RandomBrowser.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target(
+{
+  ElementType.TYPE, ElementType.METHOD
+})
+public @interface RandomBrowser
```
[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/RandomBrowser.java]: Done

File: src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/RandomBrowser.java (hunk 1 above)
Comment (oomelianchuk) [reply src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/RandomBrowser.java]: Done

File: src/main/java/com/xceptance/neodymium/module/statement/browser/BrowserStatement.java (hunk 2 above)
Comment (oomelianchuk) [reply src/main/java/com/xceptance/neodymium/module/statement/browser/BrowserStatement.java:318]: Done

File: src/test/java/com/xceptance/neodymium/testclasses/browser/RandomnessOfRandomBrowser.java (hunk 3 above)
Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/testclasses/browser/RandomnessOfRandomBrowser.java:54]: Done

File: src/main/java/com/xceptance/neodymium/module/statement/browser/BrowserStatement.java [hunk 6]
```diff
@@ -260,6 +262,8 @@ private boolean canReuse(boolean preventReuse, WebDriverStateContainer webDriver
         List<Browser> classBrowserAnnotations = findClassBrowserAnnotation(testClass.getJavaClass());
         List<SuppressBrowsers> methodSuppressBrowserAnnotations = getAnnotations(method.getMethod(), SuppressBrowsers.class);
         List<SuppressBrowsers> classSuppressBrowserAnnotations = getAnnotations(testClass.getJavaClass(), SuppressBrowsers.class);
+        List<RandomBrowser> methodRandomBrowserAnnotations = getAnnotations(method.getMethod(), RandomBrowser.class);
+        List<RandomBrowser> classRandomBrowserAnnotation = getAnnotations(testClass.getJavaClass(), RandomBrowser.class);
```
[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/main/java/com/xceptance/neodymium/module/statement/browser/BrowserStatement.java]: IMO, this annotation should be inheritable, analog to the `@Browser`. The code is now adjusted to make annotation work like this. The change can be reverted anytime, so if you have another opinion, please let me know or suggest an appointment to discuss the desired annotation behavior.

**Comment by oomelianchuk** [comment]
The annotation works as expected now. Tests for overwitting of inherited annotations are added. The key method `BrowserStatement.findBrowserRelatedClassAnnotation` is also tested now

File: src/test/java/com/xceptance/neodymium/tests/BrowserStatementTest.java [hunk 7]
```diff
@@ -96,11 +106,78 @@ public void setJUnitViewModeFlat()
...
+    public void testRandomBrowsersOverwriting()
+    {
+        // the test from RandomBrowsersOverwritingChild should be run 3 times, as the corresponding annotations from
+        // RandomBrowserParent class should be overwritten
+        Result result = JUnitCore.runClasses(RandomBrowsersOverwritingChild.class);
+        checkPass(result, 3, 0);
+    }
+
+    @Test
+    public void testFindBrowserRelatedClassAnnotationMethod()
```
Comment (occupant23) [inline src/test/java/com/xceptance/neodymium/tests/BrowserStatementTest.java:157]: This test method is producing an error:
```
java.lang.AssertionError: BrowserStatement.findBrowserRelatedClassAnnotation works not as expected expected:<3> but was:<[@com.xceptance.neodymium.module.statement.browser.multibrowser.RandomBrowsers(value=3)]>
```
The result of the function you're calling is a list of annotations. I guess you want to validate the value instead. Pleas adjust this.

File: src/test/java/com/xceptance/neodymium/testclasses/browser/mixed/MethodBrowserAnnotationOverwritesClassRandomBrowser.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+import com.xceptance.neodymium.module.statement.browser.multibrowser.Browser;
+import com.xceptance.neodymium.module.statement.browser.multibrowser.RandomBrowsers;
+
+@Browser("Chrome_1024x768")
+@Browser("Chrome_1500x1000")
+@Browser("FF_1024x768")
+@Browser("FF_1500x1000")
+@RandomBrowsers(2)
+@RunWith(NeodymiumRunner.class)
+
```
Comment (occupant23) [inline src/test/java/com/xceptance/neodymium/testclasses/browser/mixed/MethodBrowserAnnotationOverwritesClassRandomBrowser.java:16]: delete newline

File: src/test/java/com/xceptance/neodymium/tests/BrowserStatementTest.java [hunk 9]
```diff
@@ -125,6 +202,17 @@ public void testClassBrowserSuppressed() throws Throwable
...
 
+    @Test
+    public void testDisableRandomBrowserAnnotation() throws Throwable
+    {
+        //
+        String[] expected = new String[]
+        {
+          "test1"
+        };
+        checkDescription(DisableRandomBrowserAnnotation.class, expected);
```
Comment (occupant23) [inline src/test/java/com/xceptance/neodymium/tests/BrowserStatementTest.java:213]: it would be good to have a test that actually disables something ;-)
please add more test functions inside this test e.g.

```
    @RandomBrowsers(0)
    @Browser("Chrome_1024x768")
    @Browser("Chrome_1500x1000")
    @Test
    public void test2()
    {

    }
```

**Review by occupant23 — CHANGES_REQUESTED** [review]
please adjust the commented areas

File: src/test/java/com/xceptance/neodymium/tests/BrowserStatementTest.java (hunk 7 above)
Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/tests/BrowserStatementTest.java:157]: Fixed

File: src/test/java/com/xceptance/neodymium/testclasses/browser/mixed/MethodBrowserAnnotationOverwritesClassRandomBrowser.java (hunk 8 above)
Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/testclasses/browser/mixed/MethodBrowserAnnotationOverwritesClassRandomBrowser.java:16]: Removed

File: src/test/java/com/xceptance/neodymium/tests/BrowserStatementTest.java (hunk 9 above)
Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/tests/BrowserStatementTest.java:216]: test extended

File: src/test/java/com/xceptance/neodymium/testclasses/browser/DisableRandomBrowserAnnotation.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+import org.junit.runner.RunWith;
+
+import com.xceptance.neodymium.NeodymiumRunner;
+import com.xceptance.neodymium.module.statement.browser.multibrowser.Browser;
+import com.xceptance.neodymium.module.statement.browser.multibrowser.RandomBrowsers;
+
+@RunWith(NeodymiumRunner.class)
+@Browser("Chrome_1024x768")
+@Browser("Chrome_1500x1000")
+public class DisableRandomBrowserAnnotation
```
Comment (occupant23) [inline src/test/java/com/xceptance/neodymium/testclasses/browser/DisableRandomBrowserAnnotation.java:13]: please add @RandomBrowsers(1) and verify the disabling

Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/testclasses/browser/DisableRandomBrowserAnnotation.java:14]: Done

**Review by occupant23 — APPROVED** [review]
